src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/BuildPropertyPage/DefineConstantsValueProvider.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using Microsoft.VisualStudio.ProjectSystem.Debug;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Properties
+{
+    /// <summary>
+    /// Supports the setting of constants in the property page, corresponding to the "DefineConstants" property
+    /// </summary>
+    [ExportInterceptingPropertyValueProvider(DefineConstantsPropertyName, ExportInterceptingPropertyValueProviderFile.ProjectFile)]
+    [AppliesTo(ProjectCapability.VisualBasic)]
+    internal sealed class DefineConstantsValueProvider : InterceptingPropertyValueProviderBase
+    {
+        internal const string DefineConstantsPropertyName = "DefineConstants";
+
+        public override Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+        {
+            Dictionary<string, string> constantsDict = AbstractProjectConfigurationDefineConstantsEncoding.ParseIntoDictionary(unevaluatedPropertyValue);
+            return Task.FromResult<string?>(AbstractProjectConfigurationDefineConstantsEncoding.Format(constantsDict));
+        }
+
+        public override Task<string> OnGetEvaluatedPropertyValueAsync(string propertyName, string evaluatedPropertyValue, IProjectProperties defaultProperties)
+        {
+            return Task.FromResult<string>(OnGetPropertyValue(evaluatedPropertyValue));
+        }
+
+        public override Task<string> OnGetUnevaluatedPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties)
+        {
+            return Task.FromResult<string>(OnGetPropertyValue(unevaluatedPropertyValue));
+        }
+
+        private static string OnGetPropertyValue(string propertyValue)
+        {
+            Dictionary<string, string> constantsDict = AbstractProjectConfigurationDefineConstantsEncoding.ParseIntoDictionary(propertyValue);
+            return AbstractProjectConfigurationDefineConstantsEncoding.DisplayFormat(constantsDict);
```

Comment:
There are two similar snippets in this file, both calling `ParseIntoDictionary` then `*Format`.

There's no other consumer of that intermediary dictionary. In fact, I don't think it's needed.

If you use the singleton idea from earlier, you could simplify this case into:

```suggestion
            return KeyValuePairListEncoding.Instance.Santize(propertyValue);
```

Both `KeyValuePairListEncoding` and `KeyQuotedValuePairListEncoding` would define that method. To do that, you could make `AbstractProjectConfigurationDefineConstantsEncoding` an actual abstract class, and have both encoding types derive from it. The base class could define `Format` as `protected abstract` and delegate to it from its own `Sanitize` method.

I'm happy to chat about this if you like.